lib/site_settings/hidden_provider.rb

```diff
@@ -0,0 +1,19 @@
+# frozen_string_literal: true
+
+module SiteSettings
+end
+
+# A cache for providing default value based on site locale
+class SiteSettings::HiddenProvider
+  def initialize
+    @hidden_settings = []
+  end
+
+  def add_hidden(site_setting_name)
+    @hidden_settings << site_setting_name unless @hidden_settings.include?(site_setting_name)
+  end
+
+  def all
+    DiscoursePluginRegistry.apply_modifier(:hidden_site_settings, @hidden_settings)
```

Comment:
I'm wondering if we need to pass through the entire array and require the plugin to mutate the array manually when we can handle the pushing of the array into core. From the plugin perspective, the API can be as simple like: `make_hidden_site_setting(:something)`.

Reply:
Yeah I think having the consistency of `register_modifier` makes sense here 👌 

Also note we have a very similar pattern for site setting defaults

https://github.com/discourse/discourse/blob/5c38e55dc9047c121d7727352e994041d26ce952/lib/site_settings/defaults_provider.rb#L41-L41

Reply:
Sorry what I mean is that I don't think the user facing API should look something like the following because it is more complicated and harder to document.

```
register_modifier(:hidden_site_settings) do |hidden|
  (hidden + [:invite_only, :login_required]).uniq
end
```

Instead I was proposing something like this:

```
# plugin/instance.rb


make_site_setting_hidden(:something) or make_site_settings_hidden(%i{hidden})
```

The API can eventually still be implemented like this:

```
def make_site_setting_hidden(site_settings)
  register_modifier(:hidden_site_settings) do |hidden|
    hidden + site_settings
  end
end
```

> I'm not sure how make_hidden_site_setting would work in a multisite-safe way.

All sites in a multisite clusters share the same set of site settings so I'm not sure if multisite safety is a valid concern here.

Reply:
> > I'm not sure how make_hidden_site_setting would work in a multisite-safe way.
>
>All sites in a multisite clusters share the same set of site settings so I'm not sure if multisite safety is a valid concern here.

The CDCK use-case is to hide/unhide **different settings on each site** in a multisite cluster. So we want to do something like

```ruby
register_modifier(:hidden_site_settings) do |hidden|
  hidden.delete :super_advanced_setting if plan.enterprise?
  hidden << :footgun if plan.basic?
  hidden
end
```

For this kind of use-case, some kind of callback interface is essential. I don't think a `make_site_setting_hidden` API would work for this kind of use case.

Reply:
```
register_modifier(:hidden_site_settings) do |hidden|
  hidden.delete :super_advanced_setting if plan.enterprise?
  hidden << :footgun if plan.basic?
  hidden
end
```

Ahh icic this is very different from the example provided in the PR description. 

The API can then be updated to something like this:

```
make_site_setting_hidden(:something) do
  # some condition
end 
```

Personally I don't like having `register_modifier` as a public facing API for a few reasons:

1. No proper way to document how the API is used for different modifiers.
1. We allow users of the API to mutate the value being passed to the modifier in however way we want. If they mutate the values incorrect, stuff blows up.

Reply:
Moving this discussion to Dev.